opentracing-spring-cloud-starter-jaeger/src/main/java/io/opentracing/contrib/spring/cloud/starter/jaeger/JaegerAutoConfiguration.java

```diff
@@ -0,0 +1,204 @@
+import com.uber.jaeger.samplers.RemoteControlledSampler;
+import com.uber.jaeger.samplers.Sampler;
+import com.uber.jaeger.senders.Sender;
+import io.opentracing.contrib.spring.cloud.starter.jaeger.JaegerConfigurationProperties.RemoteReporterProperties;
+import io.opentracing.contrib.spring.cloud.starter.jaeger.customizers.B3CodecTracerBuilderCustomizer;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.util.StringUtils;
+
+/**
+ * @author <a href="mailto:gytis@redhat.com">Gytis Trikleris</a>
+ */
+@Configuration
+@ConditionalOnClass(com.uber.jaeger.Tracer.class)
+@ConditionalOnMissingBean(io.opentracing.Tracer.class)
+@ConditionalOnProperty(value = "opentracing.jaeger.enabled", havingValue = "true", matchIfMissing = true)
+@AutoConfigureBefore(name = "io.opentracing.contrib.spring.web.autoconfig.TracerAutoConfiguration")
+@EnableConfigurationProperties(JaegerConfigurationProperties.class)
+public class JaegerAutoConfiguration {
+
+  @Autowired(required = false)
+  private List<TracerBuilderCustomizer> tracerCustomizers = Collections.emptyList();
+
+  @Bean
+  public io.opentracing.Tracer tracer(JaegerConfigurationProperties jaegerConfigurationProperties,
+      Sampler sampler,
+      Reporter reporter) {
+
+    final Builder builder =
+        new Builder(jaegerConfigurationProperties.getServiceName())
+            .withReporter(reporter)
+            .withSampler(sampler);
+
+    tracerCustomizers.forEach(c -> c.customize(builder));
+
+    return builder.build();
+  }
+
+  @ConditionalOnMissingBean
+  @Bean
+  public Reporter reporter(JaegerConfigurationProperties properties,
+      Metrics metrics,
+      @Autowired(required = false) ReporterAppender reporterAppender) {
+    List<Reporter> reporters = new LinkedList<>();
+
+    JaegerConfigurationProperties.RemoteReporterProperties remoteReporterProperties =
+        properties.getRemoteReporterProperties();
+
+    JaegerConfigurationProperties.HttpSender httpSender = properties.getHttpSender();
+    if (!StringUtils.isEmpty(httpSender.getUrl()) && !httpSender.isDisable()) {
+      reporters.add(getHttpReporter(metrics, remoteReporterProperties, httpSender));
+    }
+
+    JaegerConfigurationProperties.UdpSender udpSender = properties.getUdpSender();
+    if (!StringUtils.isEmpty(udpSender.getHost()) && !udpSender.isDisable()) {
```

Comment:
Why would we limit the users to only one remote `Reporter`? If there isn't any technical reason from the Jaeger core perspective, I don't it makes sense to impose an artificial limitation of only one remote reporter.

Reply:
I don't know why would anybody want to report twice the same span. Actually I am not sure how the double reported span would be stored and shown in UI. Maybe it would join tags/logs.. I don't know exactly.

I think we should default to UDP sender (reporting to localhost) and use HTTP if the HTTP endpoint is specified.

Reply:
I'll go with your suggestion since you are the expert in this domain :).

Reply:
Done